canceling orphan backend

Nastradamus opened the pull request:
PostgreSQL and pgbouncer 'chain' have a problem:

If client of PgBouncer goes away in unexpected manner (for example, python client dies, network problems),
PgBouncer will send to PostgreSQL only "terminate" request (I mean terminate packet by protocol, not pg_terminate_backend).
After that, PostgreSQL will fully process
query/transaction before terminate backend.

This is a problem when we have a lot of expensive and long queries. Why should we waste CPU and I/O for 'homeless'
backends?

I created a concept of patch for resolving this problem.
This patch is for master branch.
Can anybody review it?

markokr commented:
Don't use accept_cancel_request() directly, it does lot of juggling to get server connection, you already have it.  Move the part that operates on server connection to separate function and use that.

Nastradamus commented:
But I need new server connection to send cancel packet. It will not wokr on same connection.

http://www.postgresql.org/docs/9.4/static/protocol-flow.html

"To issue a cancel request, the frontend opens a new connection to the server and sends a CancelRequest message, rather than the StartupMessage message that would ordinarily be sent across a new connection. The server will process this request and then close the connection."

markokr commented:
accept_cancel_request() already deals with it

Nastradamus commented:
In original, accept_cancel_request() is executed only in one situation: then packet header from new client connection is PKT_CANCEL. In order to avoid manipulate with socket bufer, I execute accept_cancel_request() from another place to create new server connection without real new client connection. It is simple to understand code and don't brake pgbouncer's event-driven state machine.

Nastradamus commented:
markork, could you please explain what is wrong with my code? I call accept_cancel_request() because it do all checks and manipulations that I need. No excess, unwanted checks.
Then I wrote this code, I added many logging points almost in whole 'state machine' functions. I see: it works pretty, without overhead and excess functions and manipulations with connects.

Nastradamus commented:
I added statistics column for 'SHOW STATS' (thanks to ioguix pull-request), renamed variable 'do_cancel' to 'hold_connection' and changed it to bool type.

dev1ant commented:
Any success here, guys?

markokr commented:
I have not had time to work on new features for next release.  Currently I'm still cleaning up stuff for 1.7.1.  After that is released I will look at new features.

dev1ant commented:
Cool, thanks!

dev1ant commented:
@markokr, 1.7.1 is out. Will you have time to take a look at this PR?

mcansky commented:
any update on this ? is it likely to be merged at some point ?

PJMODOS commented:
Can you not rename the usage.txt as part of this patch? Yes it could be renamed, but I don't like mixing unrelated changes like file rename with features.

PJMODOS commented:
Oh I see it's just added file by mistake and then removed, guess it needs squishing. Also the patch could use docs, the new config variable is not documented at all.

Nastradamus commented:
@markokr, do I need improve something? Or may be pgbouncer team will rewrite this patch code in more correct way?

Nastradamus commented:
Synced with current master. Squashed commits.

Nastradamus commented:
I renamed 'homeless' backends to 'orphan'.

fluancefg commented:
Hi guys,
What happens concretely with that patch ? Is it going to be merged at some point ?

NikolayS commented:
Hey @markokr @petere any thoughts/decisions regarding this PR? 2.5 years passed..

NikolayS commented:
3.5 years, woohoo!

Meanwhile: https://github.com/yandex/odyssey

anibalrivero commented:
Considering the age of this issue I guess it will never happen.
Is there a PgBouncer fork with this patch applied? I could really use it.

JelteF commented:
I like this general idea: i.e. sending a cancel request to postgres for queries for which there is no client anymore. But I'm closing this as it is too outdated to be mergeable. Feel free to open a new PR with an updated version of this.
